test_stream_zip.py

```diff
@@ -1027,27 +1046,40 @@ def extracted():
-        ('UTC+0', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=1)))),
-        ('UTC+0', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=-1)))),
-        ('UTC+1', datetime(2011, 1, 1, 1, 2, 3, 123)),
-        ('UTC+1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=0)))),
-        ('UTC+1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=1)))),
-        ('UTC+1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=-1)))),
-        ('UTC-1', datetime(2011, 1, 1, 1, 2, 3, 123)),
-        ('UTC-1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=0)))),
-        ('UTC-1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=1)))),
-        ('UTC-1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=-1)))),
+        # Datetimes near the UNIX epoch (1970)
+        ('UTC+0', datetime(1969, 12, 31, 23, 59, 58), datetime(1970,  1,  1,  0,  0,  0)),
+        ('UTC+0', datetime(1969, 12, 31, 23, 59, 59), datetime(1970,  1,  1,  0,  0,  0)),
+        ('UTC+0', datetime(1970,  1,  1,  0,  0,  0), None),
+        # Datetimes near the maximum representable datetime in the UNIX timestamp header 
+        # (4-byte signed integer counting the number of seconds since 1970)
+        ('UTC+0', datetime(2038,  1, 19,  3, 14,  7), None),
+        ('UTC+0', datetime(2038,  1, 19,  3, 14,  8), datetime(2038,  1, 19,  3, 14,  7)),
+        ('UTC+0', datetime(2038,  1, 19,  3, 14,  9), datetime(2038,  1, 19,  3, 14,  7)),
+        ('UTC+0', datetime(2038,  1, 19,  3, 14, 10), datetime(2038,  1, 19,  3, 14,  7)),
+        # Miscellaneous
+        ('UTC+0', datetime(2011, 1, 1, 1, 2, 3, 123), None),
+        ('UTC+0', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=0))), None),
+        ('UTC+0', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=1))), None),
+        ('UTC+0', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=-1))), None),
+        ('UTC+1', datetime(2011, 1, 1, 1, 2, 3, 123), None),
+        ('UTC+1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=0))), None),
+        ('UTC+1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=1))), None),
+        ('UTC+1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=-1))), None),
+        ('UTC-1', datetime(2011, 1, 1, 1, 2, 3, 123), None),
+        ('UTC-1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=0))), None),
+        ('UTC-1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=1))), None),
+        ('UTC-1', datetime(2011, 1, 1, 1, 2, 3, 123, tzinfo=timezone(timedelta(hours=-1))), None),
     ],
 )
-def test_unzip_modification_time(method, timezone, modified_at):
+def test_unzip_modification_time(method, timezone, modified_at, expected_modified_at):
```

Comment:
I'm just running the tests locally, and I'm getting some errors... I wonder if it's because I'm currently in British Summer Time, 1 our off from UTC...

```bash
pytest 'test_stream_zip.py::test_unzip_modification_time[UTC+0-modified_at0-expected_modified_at0-ZIP_32]'
```
Results in 
```bash
>           assert os.path.getmtime('my_file') == int(expected_modified_at.timestamp())
E           AssertionError: assert 0.0 == -3600
E            +  where 0.0 = <function getmtime at 0x101338040>('my_file')
E            +    where <function getmtime at 0x101338040> = <module 'posixpath' (frozen)>.getmtime
E            +      where <module 'posixpath' (frozen)> = os.path
E            +  and   -3600 = int(-3600.0)
E            +    where -3600.0 = <built-in method timestamp of datetime.datetime object at 0x1040fb990>()
E            +      where <built-in method timestamp of datetime.datetime object at 0x1040fb990> = datetime.datetime(1970, 1, 1, 0, 0).timestamp
```
The tests `test_unzip_modification_time` do pass for me. 

Not quite sure why right now...

Reply:
According to [this jszip issue](https://github.com/Stuk/jszip/issues/369), convention is that MS-DOS timestamps are in local time, whereas extended timestamps use UNIX conventions which means all timestamps are values in UTC.

Both `os.path.getmtime()` and `datetime.timestamp()` seem to respect that.

```python
# These produce the same value
datetime.datetime.now().timestamp()
datetime.datetime.now(datetime.UTC).timestamp()

# as do these
datetime.datetime.fromtimestamp(os.path.getmtime('some-file')).timestamp()
datetime.datetime.fromtimestamp(os.path.getmtime('some-file'), datetime.UTC).timestamp()
```

But a `datetime.datetime(1970, 1, 1, 0, 0)` with no timezone information will be interpreted as a local time, so the timestamp will have a timezone offset. IOW, here in NYC at UTC-5:

```python
>>> dt = datetime.datetime(1970, 1, 1, 0, 0, 0, tzinfo=None)
>>> dt.timestamp()
18000.0
>>> dtutc = datetime.datetime(1970, 1, 1, 0, 0, 0, tzinfo=datetime.UTC)
>>> dtutc.timestamp()
0.0
```